source/MainWindow.cpp

```diff
@@ -263,38 +264,18 @@ void MainWindow::MessageReceived(BMessage *msg)
 			break;
 		}
 		case M_SET_TILE_COUNT_3:
-		{
-			fGridSize = 3;
-			gPreferences.ReplaceInt8("gridsize",3);
-			GenerateGrid(fGridSize);
-			break;
-		}
 		case M_SET_TILE_COUNT_4:
-		{
-			fGridSize = 4;
-			gPreferences.ReplaceInt8("gridsize",4);
-			GenerateGrid(fGridSize);
-			break;
-		}
 		case M_SET_TILE_COUNT_5:
-		{
-			fGridSize = 5;
-			gPreferences.ReplaceInt8("gridsize",5);
-			GenerateGrid(fGridSize);
-			break;
-		}
 		case M_SET_TILE_COUNT_6:
-		{
-			fGridSize = 6;
-			gPreferences.ReplaceInt8("gridsize",6);
-			GenerateGrid(fGridSize);
-			break;
-		}
 		case M_SET_TILE_COUNT_7:
 		{
-			fGridSize = 7;
-			gPreferences.ReplaceInt8("gridsize",7);
-			GenerateGrid(fGridSize);
+			uint8 gridSize = msg->what - M_SET_TILE_COUNT_3 + 3;
```

Comment:
This is not very obvious.
I would use a single M_SET_TILE_COUNT message and add a "gridsize" field to it instead.

I also wonder if changing the grid size should generate a new game immediately. If you already started playing, it would be not so nice to erase your work, so either we should add an alert, or in that case the change should only happen when you start a new game?

Reply:
M_SET_TILE_COUNT_3 and so on are actually triggered by (and only by) the user selecting a different difficulty level. The fGridSize used in the original code actually means grid count, which corresponds to difficulty levels. It is totally unrelated to Tile Size the user selects, which won't trigger a new game. (Please see issue #12.)

Reply:
I see the difference with tile size (which is just a change in how we display things).

So, here is the scenario in issue #12:
- User starts the application. Timer starts counting.
- User changes difficulty level
- A new game is immediately started (with the new difficulty level), but the timer keeps ticking from the previous game

Your changes will fix this, but what about:

- User starts the app
- User starts to lay out tiles
- User changes difficulty level
- A new game starts, timer is reset, and the grid is empty. All the hard work playing the current game is lost

I think this is unexpected, and I would consider either:
- Adding an alert asking the user if they want to lose the current game
- Or, making the change only apply for the next game, leaving the current one unaffected

I agree that this goes further than what is raised in issue #12, but it would make sense to rework the code this way, still.

Reply:
> I think this is unexpected, and I would consider either:
> * Adding an alert asking the user if they want to lose the current game
> * Or, making the change only apply for the next game, leaving the current one unaffected

I'd prefer the alert. If nothing happens until the next game is started, the user assumes the menu just doesn't work. "Number Base" has the same issue, BTW.